### PR TITLE
create a C++ EOS interface

### PR DESCRIPTION
**Pull request by zingale**

### CODE REVIEW

**Comment by zingale** [comment]
blocked by #247

**Comment by zingale** [comment]
`test_eos_C` doesn't compile for me -- it cannot find `eos_type.H`

**Comment by maxpkatz** [comment]
> `test_eos_C` doesn't compile for me -- it cannot find `eos_type.H`

This is fixed, I had forgotten to check that new file in (also eos_override.H).

**Comment by zingale** [comment]
Now I get:
```
../../interfaces/eos.H:205:7: error: ‘print_state’ was not declared in this scope; did you mean ‘initstate’?
  205 |       print_state(state);
      |       ^~~~~~~~~~~
```

**Comment by maxpkatz** [comment]
> Now I get:
> 
> ```
> ../../interfaces/eos.H:205:7: error: ‘print_state’ was not declared in this scope; did you mean ‘initstate’?
>   205 |       print_state(state);
>       |       ^~~~~~~~~~~
> ```

That should be fixed now; we have to declare functions before we use them (and in this case, we also just define them immediately).

**Comment by zingale** [comment]
okay, it works now and gives the same answers as before the header-only changes.

File: EOS/eos_composition.H [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  // Calculate abar, the mean nucleon number,
+  // zbar, the mean proton number,
+  // mu, the mean molecular weight,
+  // mu_e, the mean number of nucleons per electron, and
+  // y_e, the electron fraction.
+
+  Real sum = 0;
+  for (int n = 0; n < NumSpec; n++) {
+    sum += state.xn[n] * zion[n] / aion[n];
+  }
```
Comment (zingale) [inline EOS/eos_composition.H]: aion_inv is the the header

File: EOS/eos_composition.H [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  state.mu_e = 1.0 / sum;
+  state.y_e = 1.0 / state.mu_e;
+
+  sum = 0;
+
+  for (int n = 0; n < NumSpec; n++) {
+    sum += state.xn[n] / aion[n];
+  }
```
Comment (zingale) [inline EOS/eos_composition.H]: aion_inv

File: EOS/eos_composition.H [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+eos_xderivs_t composition_derivatives(eos_t& state) {
+
+  eos_xderivs_t state_xderivs;
+
+#ifdef EXTRA_THERMO
+  for (int n = 0; n < NumSpec; n++) {
+    state_xderivs.dpdX[n] =
+      state.dpdA * (state.abar / aion[n]) * (aion[n] - state.abar) +
+      state.dpdZ * (state.abar / aion[n]) * (zion[n] - state.zbar);
+
```
Comment (zingale) [inline EOS/eos_composition.H]: aion_inv

File: EOS/gamma_law_general/actual_eos.H [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+  const Real m_nucleon = 1.0 / n_A;
+  const Real fac = 1.0 / std::pow(2.0 * M_PI * hbar * hbar, 1.5);
+
+  if (eos_assume_neutral) {
+    state.mu = state.abar;
+
+  } else {
+    Real sum = 0.0;
+    for (int n = 0; n < NumSpec; n++) {
+      sum += (zion[n] + 1.0) * state.xn[n] / aion[n];
```
Comment (zingale) [inline EOS/gamma_law_general/actual_eos.H]: aion_inv

File: EOS/gamma_law_general/actual_eos.H [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+      break;
+
+    case eos_input_rh:
+
+      // dens, enthalpy, and xmass are inputs
+
+      // Solve for the temperature:
+      // h = e + p/rho = (p/rho)*[1 + 1/(gamma-1)] = (p/rho)*gamma/(gamma-1)
+
+      state.T = (state.h * state.mu * m_nucleon / k_B)*(eos_gamma - 1.0)/eos_gamma;
```
Comment (zingale) [inline EOS/gamma_law_general/actual_eos.H]: I think we should store m_nucleon / k_B as a constant, since that appears so frequently together.

Comment (maxpkatz) [reply EOS/gamma_law_general/actual_eos.H]: We could do that. Alternatively if we group this as `(m_nucleon / k_B)` then the compiler should evaluate it at compile time.

**Pull request merged**